packages/clay-modal/src/ClayModal.tsx

```diff
@@ -0,0 +1,206 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import Button from '@clayui/button';
+import classNames from 'classnames';
+import Icon from '@clayui/icon';
+import React, {FunctionComponent, useEffect, useRef, useState} from 'react';
+import {createPortal} from 'react-dom';
+import {Footer, FooterContainer, FooterContainerProps} from './Footer';
+import {useModal} from './Hook';
+
+type Size = 'full-screen' | 'lg' | 'sm';
+
+type Status = 'danger' | 'info' | 'success' | 'warning';
+
+interface Props extends React.HTMLAttributes<HTMLDivElement> {
+	/**
+	 * Callback called to close the modal.
+	 */
+	onClose: () => void;
+
+	/**
+	 * Renders the footer with the FooterContainer component utilities.
+	 */
+	renderFooter?: React.ReactElement;
+
+	/**
+	 * The size of element modal.
+	 */
+	size?: Size;
+
+	/**
+	 * The path to the SVG spritemap file containing the icons.
+	 */
+	spritemap?: string;
+
+	/**
+	 * Status messages.
+	 */
+	status?: Status;
+
+	/**
+	 * Title of the modal.
+	 */
+	title?: string;
+
+	/**
+	 * Url to place an iframe in the body of the modal.
+	 */
+	url?: string;
+
+	/**
+	 * Flag to indicate if the modal is open or closed.
+	 */
+	visible: boolean;
+}
+
+const ICON_MAP = {
+	danger: 'exclamation-full',
+	info: 'info-circle',
+	success: 'check-circle',
+	warning: 'question-circle-full',
+};
+
+const Portal = ({children}: {children: React.ReactNode}) => {
+	const portalRef = useRef(document.createElement('div'));
+	const elToMountTo = document && document.body;
+
+	useEffect(() => {
+		elToMountTo.appendChild(portalRef.current);
+		return () => {
+			elToMountTo.removeChild(portalRef.current);
+		};
+	}, [elToMountTo]);
+
+	return createPortal(children, portalRef.current);
+};
```

Comment:
In theory there should be only one modal on the screen at a time, but the code here doesn't seem to do anything to enforce that. You can instantiate multiple modals, and each will append a new div and create a portal.

Yet further down, we are adding and removing classes to/from the `document.body`, which is effectively a singleton. So we have an impedance mismatch, with some parts of the code behaving like a multiple instantiation is ok, and others behaving like they are operating inside a singleton. We should probably look at fixing that.

Reply:
I'm not sure if we would have a case of opening two modal at the same time, otherwise, we would have no problem with manipulating the `document.body`.

I'm going to improve this from the `backdrop` Portal, but I do not know if we can control Modal content as a singleton. Maybe dev will have to handle this in your application. I was thinking of creating some utility that the dev could call via context to manipulate the modal in its application.

Reply:
> I'm not sure if we would have a case of opening two modal at the same time

While we are responsible for how we use it in Portal, as an open source library, there are no guarantees about how others will do so. Ideally, we'd have something that is "fool-proof", and my comment was trying to get at the fact that it currently is not fool-proof (because it assumes singleton semantics without actually enforcing them).

Reply:
Yeah, you're right about that.